test/e2e/tests/vision/vision.spec.ts

```diff
@@ -0,0 +1,144 @@
+    // Assert that the text was correctly inserted
+    await expect(queryEditor).toHaveText(inputText)
+
+    const paramsInputText = JSON.stringify({id: bookDocument._id})
+    // Type text into the params editor
+    await paramsEditor.fill(paramsInputText.slice(0, -2))
+    // Error icon should be visible
+    await expect(paramsRegion.locator('[data-sanity-icon="error-outline"]')).toBeVisible()
+
+    // Fill the params editor with the correct text
+    await paramsEditor.fill(paramsInputText)
+    // Error icon should not be visible
+    await expect(paramsRegion.locator('[data-sanity-icon="error-outline"]')).not.toBeVisible()
+    // Assert that the text was correctly inserted
+    await expect(paramsEditor).toHaveText(paramsInputText)
+
+    // Find the button with the text "Fetch" and click it.
+    await page.locator('button').filter({hasText: 'Fetch'}).click()
+
+    // Assert that the results are visible
+    // It should find the book document assert that by checking the title and the id
+    await expect(resultRegion.getByText(bookTitle)).toBeVisible()
+    await expect(resultRegion.getByText(bookDocument._id)).toBeVisible()
+  })
+
+  test('should be possible to paste and parse a query', async ({
+    page,
+    context,
+    sanityClient,
+    browserName,
+  }) => {
+    // Firefox doesn't support pasting from the clipboard
+    test.skip(browserName === 'firefox')
+
+    const bookTitle = 'Test Book'
+    const bookDocument = await sanityClient.create({
+      _type: 'book',
+      title: bookTitle,
+    })
+
+    // Grant clipboard permissions before opening the page
+    await context.grantPermissions(['clipboard-read', 'clipboard-write'])
+
+    await openVisionTool(page)
+    const query = `*[_type == "book" && _id == $id]{_id, title}`
+    const params = {id: bookDocument._id}
+    const url = sanityClient.getUrl(
+      sanityClient.getDataUrl('query', encodeQueryString(query, params)),
+    )
+    await page.evaluate((text) => {
+      navigator.clipboard.writeText(text)
+    }, url)
+
+    const {queryEditor, paramsEditor, resultRegion, queryEditorRegion} =
+      await getVisionRegions(page)
+    await queryEditorRegion.click()
+    await queryEditor.focus()
+    // // Paste the url into the query editor
+    await page.keyboard.press('ControlOrMeta+V')
+
+    // Assert that the text was correctly inserted
+    await expect(queryEditor).toHaveText(query)
+
+    const paramsText = await paramsEditor.textContent()
+    const parsedParams = JSON.parse(paramsText || '{}')
+    expect(parsedParams).toHaveProperty('id', bookDocument._id)
+
+    // The query executes automatically when a url is pasted, so it should have results
+    // Assert that the results are visible
+    // It should find the book document assert that by checking the title and the id
+    await expect(resultRegion.getByText(bookTitle)).toBeVisible({
+      timeout: 10_000,
+    })
+    await expect(resultRegion.getByText(bookDocument._id)).toBeVisible()
```

Comment:
We should avoid relying on timeouts, sometimes they are needed but it gives me the feeling we could probably be doing this in a different way that wouldn'rt require it.

The same comment about the expectations remains as on a previous comment :)